--- a/net/minecraft/server/network/PlayerChunkSender.java
+++ b/net/minecraft/server/network/PlayerChunkSender.java
@@ -44,6 +_,11 @@
     public void dropChunk(ServerPlayer player, ChunkPos chunkPos) {
         if (!this.pendingChunks.remove(chunkPos.toLong()) && player.isAlive()) {
-            player.connection.send(new ClientboundForgetLevelChunkPacket(chunkPos));
+            org.ulyxspigot.ulyxspigot.async.UlyxAsyncChunkSending.send(player.connection, new ClientboundForgetLevelChunkPacket(chunkPos));
+            // Paper start - PlayerChunkUnloadEvent
+            if (io.papermc.paper.event.packet.PlayerChunkUnloadEvent.getHandlerList().getRegisteredListeners().length > 0) {
+                new io.papermc.paper.event.packet.PlayerChunkUnloadEvent(player.getBukkitEntity().getWorld().getChunkAt(chunkPos.longKey), player.getBukkitEntity()).callEvent();
+            }
+            // Paper end - PlayerChunkUnloadEvent
         }
     }
 
@@ -64,13 +_,13 @@
                     if (!list.isEmpty()) {
                         ServerGamePacketListenerImpl serverGamePacketListenerImpl = player.connection;
                         this.unacknowledgedBatches++;
-                        serverGamePacketListenerImpl.send(ClientboundChunkBatchStartPacket.INSTANCE);
+                        org.ulyxspigot.ulyxspigot.async.UlyxAsyncChunkSending.send(serverGamePacketListenerImpl, ClientboundChunkBatchStartPacket.INSTANCE);
 
                         for (LevelChunk levelChunk : list) {
                             sendChunk(serverGamePacketListenerImpl, serverLevel, levelChunk);
                         }
 
-                        serverGamePacketListenerImpl.send(new ClientboundChunkBatchFinishedPacket(list.size()));
+                        org.ulyxspigot.ulyxspigot.async.UlyxAsyncChunkSending.send(serverGamePacketListenerImpl, new ClientboundChunkBatchFinishedPacket(list.size()));
                         this.batchQuota = this.batchQuota - list.size();
                     }
                 }
