From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: UlyxSpigot <dev@ulyxspigot.local>
Date: Thu, 19 Mar 2026 21:30:00 +0700
Subject: [PATCH] UlyxSpigot brand display config

diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 8588b380f9b5375cdb2515a01e46a5731ad3ab33..9999999999999999999999999999999999999999 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2678,7 +2678,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return org.ulyxspigot.ulyxspigot.UlyxConfig.getServerBrandNameDisplay(); // UlyxSpigot - configurable brand display
     }
 
     public SystemReport fillSystemReport(SystemReport report) {
