From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sat, 1 Feb 2025 19:24:47 +0300
Subject: [PATCH] Snowball and Egg knockback


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index 448f9564abad3781b10525949fb0ba5413a87cd4..1ea3b5fdcb9ce48e7ba0fe31dd39cab1c8ab1282 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -55,6 +55,12 @@ public class Snowball extends ThrowableItemProjectile {
         Entity entity = result.getEntity();
         int i = entity.level().purpurConfig.snowballDamage >= 0 ? entity.level().purpurConfig.snowballDamage : entity instanceof Blaze ? 3 : 0; // Purpur - Add configurable snowball damage
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), i);
+        // DivineMC start - Make snowball can knockback player
+        if (this.level().divineConfig.snowballCanKnockback && entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            serverPlayer.knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ());
+        }
+        // DivineMC end - Make snowball can knockback player
     }
 
     // Purpur start - options to extinguish fire blocks with snowballs - borrowed and modified code from ThrownPotion#onHitBlock and ThrownPotion#dowseFire
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
index c7db74ccc3f8c46e97c24857f297fe5fb5f45e36..13c91d42a6ca4f2b4baa66a8ca73a391bf128ce7 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
@@ -55,7 +55,14 @@ public class ThrownEgg extends ThrowableItemProjectile {
     @Override
     protected void onHitEntity(EntityHitResult result) {
         super.onHitEntity(result);
+        net.minecraft.world.entity.Entity entity = result.getEntity(); // DivineMC - make egg can knockback player
         result.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // DivineMC start - Make egg can knockback player
+        if (this.level().divineConfig.eggCanKnockback && entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            serverPlayer.knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ());
+        }
+        // DivineMC end - Make egg can knockback player
     }
 
     @Override
