From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sun, 23 Mar 2025 01:16:35 +0300
Subject: [PATCH] Dynamic Activation of Brain


diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index bc62da127a01b33ce8fc0f958bc232d6287d57ec..9687a31e83d2d5be833cdd127ccef1415f3a3406 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -167,6 +167,21 @@ public final class ActivationRange {
                 }
 
                 ActivationRange.activateEntity(entity);
+
+                // DivineMC start - Dynamic Activation of Brain
+                if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.dabEnabled && entity.getType().dabEnabled && (!org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.dabDontEnableIfInWater || entity.getType().is(net.minecraft.tags.EntityTypeTags.CAN_BREATHE_UNDER_WATER) || !entity.isInWaterOrRain())) {
+                    if (!entity.activatedPriorityReset) {
+                        entity.activatedPriorityReset = true;
+                        entity.activatedPriority = org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.dabMaximumActivationFrequency;
+                    }
+                    int squaredDistance = (int) player.distanceToSqr(entity);
+                    entity.activatedPriority = squaredDistance > org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.dabStartDistanceSquared ?
+                            Math.max(1, Math.min(squaredDistance >> org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.dabActivationDistanceMod, entity.activatedPriority)) :
+                            1;
+                } else {
+                    entity.activatedPriority = 1;
+                }
+                // DivineMC end - Dynamic Activation of Brain
             }
         }
     }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index c87ebe4c3d69c2d3df527f9a6803d0aeb483872b..9e726a2cfb7bcf0a9a5f18e41df2ed579eae5e1b 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -862,6 +862,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             this.entityTickList
                 .forEach(
                     entity -> {
+                        entity.activatedPriorityReset = false; // DivineMC - Dynamic Activation of Brain
                         if (!entity.isRemoved()) {
                             if (!tickRateManager.isEntityFrozen(entity)) {
                                 profilerFiller.push("checkDespawn");
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 427de0ac0164bf60d3e477cb87a9255275c7aa65..6a9498766856a7c4f0fa209bcaadac70849ad11b 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -369,6 +369,8 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     public boolean fixedPose = false; // Paper - Expand Pose API
     private final int despawnTime; // Paper - entity despawn time limit
     public int totalEntityAge; // Paper - age-like counter for all entities
+    public boolean activatedPriorityReset = false; // DivineMC - Dynamic Activation of Brain
+    public int activatedPriority = org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.dabMaximumActivationFrequency; // DivineMC - Dynamic Activation of Brain
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
     // Paper start - EAR 2
     public final boolean defaultActivationState;
diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index f0731609fbfb06ea23baba2a8b3694003a470b7d..1ccab891e10f2a79435c12a9e58f5f3acc804323 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1220,6 +1220,7 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private final boolean canSpawnFarFromPlayer;
     private final int clientTrackingRange;
     private final int updateInterval;
+    public boolean dabEnabled = false; // DivineMC - Dynamic Activation of Brain
     private final String descriptionId;
     private @Nullable Component description;
     private final Optional<ResourceKey<LootTable>> lootTable;
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 09016282c0c6850d19e48966426d480310726a73..137c58d407e2a1f3923b674d5724685a3f4a697f 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -222,10 +222,10 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     @Override
     public void inactiveTick() {
         super.inactiveTick();
-        if (this.goalSelector.inactiveTick()) {
+        if (this.goalSelector.inactiveTick(this.activatedPriority, true)) { // DivineMC - Dynamic Activation of Brain
             this.goalSelector.tick();
         }
-        if (this.targetSelector.inactiveTick()) {
+        if (this.targetSelector.inactiveTick(this.activatedPriority, true)) { // DivineMC - Dynamic Activation of Brain
             this.targetSelector.tick();
         }
     }
@@ -832,17 +832,17 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         int i = this.tickCount + this.getId();
         if (i % 2 != 0 && this.tickCount > 1) {
             profilerFiller.push("targetSelector");
-            this.targetSelector.tickRunningGoals(false);
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) this.targetSelector.tickRunningGoals(false); // DivineMC - Dynamic Activation of Brain
             profilerFiller.pop();
             profilerFiller.push("goalSelector");
-            this.goalSelector.tickRunningGoals(false);
+            if (this.goalSelector.inactiveTick(this.activatedPriority, false)) this.goalSelector.tickRunningGoals(false); // DivineMC - Dynamic Activation of Brain
             profilerFiller.pop();
         } else {
             profilerFiller.push("targetSelector");
-            this.targetSelector.tick();
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) this.targetSelector.tick(); // DivineMC - Dynamic Activation of Brain
             profilerFiller.pop();
             profilerFiller.push("goalSelector");
-            this.goalSelector.tick();
+            if (this.goalSelector.inactiveTick(this.activatedPriority, false)) this.goalSelector.tick(); // DivineMC - Dynamic Activation of Brain
             profilerFiller.pop();
         }
 
diff --git a/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java b/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
index 389f773e22872778fc434ad47f3a983b490a0284..4d0d7ac936307e399fccb3e6815248bb9345ff86 100644
--- a/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
@@ -36,7 +36,11 @@ public class VillagerPanicTrigger extends Behavior<Villager> {
 
     @Override
     protected void tick(ServerLevel level, Villager owner, long gameTime) {
-        if (gameTime % 100L == 0L) {
+        // DivineMC start - Dynamic Activation of Brain
+        if (owner.nextGolemPanic < 0) owner.nextGolemPanic = gameTime + 100;
+        if (--owner.nextGolemPanic < gameTime) {
+            owner.nextGolemPanic = -1;
+        // DivineMC end - Dynamic Activation of Brain
             owner.spawnGolemIfNeeded(level, gameTime, 3);
         }
     }
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 859b859d29b637200cf7c9a0bd52d9f712413e3d..f2ad27034919add0b73822844c4650a67bcc40ee 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -36,10 +36,14 @@ public class GoalSelector {
     }
 
     // Paper start - EAR 2
-    public boolean inactiveTick() {
+    // DivineMC start - Dynamic Activation of Brain
+    public boolean inactiveTick(int tickRate, boolean inactive) {
+        if (inactive && !org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.dabEnabled) tickRate = 4;
+        tickRate = Math.min(tickRate, 3);
         this.curRate++;
-        return this.curRate % 3 == 0; // TODO newGoalRate was already unused in 1.20.4, check if this is correct
+        return this.curRate % tickRate == 0;
     }
+    // DivineMC end - Dynamic Activation of Brain
 
     public boolean hasTasks() {
         for (WrappedGoal task : this.availableGoals) {
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 8635ecfa49e2504c981dd48acbfbca5303a37611..d054c507673cf0f3077d1775b7f509cab05263d5 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -114,6 +114,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     private float spinningAnimationTicks0;
     public boolean forceDancing = false; // CraftBukkit
     private org.purpurmc.purpur.controller.FlyingMoveControllerWASD purpurController; // Purpur - Ridables
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Allay(EntityType<? extends Allay> type, Level level) {
         super(type, level);
@@ -277,13 +278,16 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("allayBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("allayActivityUpdate");
-        AllayAi.updateActivity(this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("allayActivityUpdate");
+            AllayAi.updateActivity(this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override
diff --git a/net/minecraft/world/entity/animal/axolotl/Axolotl.java b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
index e5ac43bf5631254c31d91abb1c5de408807429dd..d781e6039bd0c5bda84e01137a1a028a6c58b6d9 100644
--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -113,6 +113,7 @@ public class Axolotl extends Animal implements Bucketable {
     public final BinaryAnimator onGroundAnimator = new BinaryAnimator(10, EasingType.IN_OUT_SINE);
     public final BinaryAnimator movingAnimator = new BinaryAnimator(10, EasingType.IN_OUT_SINE);
     private static final int REGEN_BUFF_BASE_DURATION = 100;
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Axolotl(EntityType<? extends Axolotl> type, Level level) {
         super(type, level);
@@ -374,16 +375,19 @@ public class Axolotl extends Animal implements Bucketable {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("axolotlBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("axolotlActivityUpdate");
-        AxolotlAi.updateActivity(this);
-        profilerFiller.pop();
-        if (!this.isNoAi()) {
-            Optional<Integer> memory = this.getBrain().getMemory(MemoryModuleType.PLAY_DEAD_TICKS);
-            this.setPlayingDead(memory.isPresent() && memory.get() > 0);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("axolotlActivityUpdate");
+            AxolotlAi.updateActivity(this);
+            profilerFiller.pop();
+            if (!this.isNoAi()) {
+                Optional<Integer> memory = this.getBrain().getMemory(MemoryModuleType.PLAY_DEAD_TICKS);
+                this.setPlayingDead(memory.isPresent() && memory.get() > 0);
+            }
         }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     public static AttributeSupplier.Builder createAttributes() {
diff --git a/net/minecraft/world/entity/animal/frog/Frog.java b/net/minecraft/world/entity/animal/frog/Frog.java
index 7d6dbdef68b25e774b5a83bb8aeb2535246459aa..f87a7fbcca28b28369f5425e7ff5876aae0f3514 100644
--- a/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/net/minecraft/world/entity/animal/frog/Frog.java
@@ -107,6 +107,7 @@ public class Frog extends Animal {
     public final AnimationState swimIdleAnimationState = new AnimationState();
     private org.purpurmc.purpur.controller.MoveControllerWASD purpurLandController; // Purpur - Ridables
     private org.purpurmc.purpur.controller.WaterMoveControllerWASD purpurWaterController; // Purpur - Ridables
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Frog(EntityType<? extends Animal> type, Level level) {
         super(type, level);
@@ -261,13 +262,16 @@ public class Frog extends Animal {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("frogBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("frogActivityUpdate");
-        FrogAi.updateActivity(this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("frogActivityUpdate");
+            FrogAi.updateActivity(this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override
diff --git a/net/minecraft/world/entity/animal/frog/Tadpole.java b/net/minecraft/world/entity/animal/frog/Tadpole.java
index ae65292c9be65894532f11833c68e68970dbb257..946738a16dc510e055904b8d6b033801c882d003 100644
--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -66,6 +66,7 @@ public class Tadpole extends AbstractFish {
     );
     public boolean ageLocked; // Paper
     private org.purpurmc.purpur.controller.WaterMoveControllerWASD purpurController; // Purpur - Ridables
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Tadpole(EntityType<? extends AbstractFish> type, Level level) {
         super(type, level);
@@ -138,13 +139,16 @@ public class Tadpole extends AbstractFish {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("tadpoleBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("tadpoleActivityUpdate");
-        TadpoleAi.updateActivity(this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("tadpoleActivityUpdate");
+            TadpoleAi.updateActivity(this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     public static AttributeSupplier.Builder createAttributes() {
diff --git a/net/minecraft/world/entity/animal/goat/Goat.java b/net/minecraft/world/entity/animal/goat/Goat.java
index f4ad672f7cb33bd95ca2dab9ac8637b43eacf0f6..1453bcf553a5c1ffaef21d8ed1e4ece156527868 100644
--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -96,6 +96,7 @@ public class Goat extends Animal {
     private static final boolean DEFAULT_HAS_RIGHT_HORN = true;
     private boolean isLoweringHead;
     private int lowerHeadTick;
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Goat(EntityType<? extends Goat> type, Level level) {
         super(type, level);
@@ -236,13 +237,16 @@ public class Goat extends Animal {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("goatBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("goatActivityUpdate");
-        GoatAi.updateActivity(this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("goatActivityUpdate");
+            GoatAi.updateActivity(this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/hoglin/Hoglin.java b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
index 460640f5512be6553f1ec69e20e27c7cc797068b..1c9805eac0ab6f3c13490cc7b8a992c2bb6028d0 100644
--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -91,6 +91,7 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
         MemoryModuleType.PACIFIED,
         MemoryModuleType.IS_PANICKING
     );
+    private int behaviorTick; // DivineMC - Dynamic Activation of Brain
 
     public Hoglin(EntityType<? extends Hoglin> type, Level level) {
         super(type, level);
@@ -211,19 +212,22 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("hoglinBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        HoglinAi.updateActivity(this);
-        if (this.isConverting()) {
-            this.timeInOverworld++;
-            if (this.timeInOverworld > 300) {
-                this.makeSound(SoundEvents.HOGLIN_CONVERTED_TO_ZOMBIFIED);
-                this.finishConversion();
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            HoglinAi.updateActivity(this);
+            if (this.isConverting()) {
+                this.timeInOverworld++;
+                if (this.timeInOverworld > 300) {
+                    this.makeSound(SoundEvents.HOGLIN_CONVERTED_TO_ZOMBIFIED);
+                    this.finishConversion();
+                }
+            } else {
+                this.timeInOverworld = 0;
             }
-        } else {
-            this.timeInOverworld = 0;
         }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index 6f1c3d423b7a37aa261c78a0c8f5e1cb02eee320..effdcdb629579104ad48b85438dd50e6dc34e401 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -135,6 +135,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
     private static final com.mojang.serialization.Codec<java.util.Set<net.minecraft.world.item.Item>> ITEM_SET_CODEC = net.minecraft.core.registries.BuiltInRegistries.ITEM
         .byNameCodec().listOf().xmap(java.util.HashSet::new, List::copyOf);
     // CraftBukkit end
+    private int behaviorTick; // DivineMC - Dynamic Activation of Brain
 
     public Piglin(EntityType<? extends AbstractPiglin> type, Level level) {
         super(type, level);
@@ -365,11 +366,14 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("piglinBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        PiglinAi.updateActivity(this);
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            PiglinAi.updateActivity(this);
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override
diff --git a/net/minecraft/world/entity/monster/warden/Warden.java b/net/minecraft/world/entity/monster/warden/Warden.java
index 4f4bf16cfa1390f19a7e8745d86e42e24ffb3bbc..42535a9dd42e705c9d33d21038bbdbbf27272319 100644
--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -110,6 +110,7 @@ public class Warden extends Monster implements VibrationSystem {
     private final VibrationSystem.User vibrationUser;
     private VibrationSystem.Data vibrationData;
     AngerManagement angerManagement = new AngerManagement(this::canTargetEntity, Collections.emptyList());
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Warden(EntityType<? extends Monster> type, Level level) {
         super(type, level);
@@ -304,20 +305,23 @@ public class Warden extends Monster implements VibrationSystem {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("wardenBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
-        if ((this.tickCount + this.getId()) % 120 == 0) {
-            applyDarknessAround(level, this.position(), this, 20);
-        }
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+            if ((this.tickCount + this.getId()) % 120 == 0) {
+                applyDarknessAround(level, this.position(), this, 20);
+            }
 
-        if (this.tickCount % 20 == 0) {
-            this.angerManagement.tick(level, this::canTargetEntity);
-            this.syncClientAngerLevel();
-        }
+            if (this.tickCount % 20 == 0) {
+                this.angerManagement.tick(level, this::canTargetEntity);
+                this.syncClientAngerLevel();
+            }
 
-        WardenAi.updateActivity(this);
+            WardenAi.updateActivity(this);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 0c110421074d57e7283168c1133cd4ee314b57f6..47d5affe3c903318a4279f57807388446a74654e 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -181,6 +181,8 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
     );
     private boolean isLobotomized = false; public boolean isLobotomized() { return this.isLobotomized; } // Purpur - Lobotomize stuck villagers
     private int notLobotomizedCount = 0; // Purpur - Lobotomize stuck villagers
+    public long nextGolemPanic = -1; // DivineMC - Dynamic Activation of Brain
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Villager(EntityType<? extends Villager> type, Level level) {
         this(type, level, VillagerType.PLAINS);
@@ -410,7 +412,7 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         } else {
             this.isLobotomized = false;
         }
-        if (!inactive && (getRider() == null || !this.isControllable())) { // Purpur - Ridables
+        if (!inactive && this.behaviorTick++ % this.activatedPriority == 0 && (getRider() == null || !this.isControllable())) { // Purpur - Ridables // DivineMC - Dynamic Activation of Brain
             this.getBrain().tick(level, this); // Paper - EAR 2
         }
         else if (this.isLobotomized && shouldRestock(level)) restock();
