From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sat, 1 Feb 2025 18:55:59 +0300
Subject: [PATCH] MSPT Tracking for each world


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 1b7f94ea60d3fe81b4eb52fb67d2949a7e3bd2c2..c5a1bf215aa91c2dc67926385fcb93f5e0cc3549 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1814,7 +1814,15 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // DivineMC start - Parallel world ticking
     private void tickLevel(ServerLevel serverLevel, BooleanSupplier hasTimeLeft) {
         try {
+            // DivineMC start - MSPT Tracking for each world
+            long i = Util.getNanos();
             serverLevel.tick(hasTimeLeft);
+            long j = Util.getNanos() - i;
+
+            serverLevel.tickTimes5s.add(this.tickCount, j);
+            serverLevel.tickTimes10s.add(this.tickCount, j);
+            serverLevel.tickTimes60s.add(this.tickCount, j);
+            // DivineMC end - MSPT Tracking for each world
         } catch (Throwable levelTickingException) {
             CrashReport crashReport = CrashReport.forThrowable(levelTickingException, "Exception ticking world");
             serverLevel.fillReportDetails(crashReport);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 0dce7fa411479794d1586c4049c3d92abbe632d2..fcc172a2aeee3fe6bfd526411e1e123d0f80c3e3 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -604,6 +604,12 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
     // Paper end - chunk tick iteration
 
+    // DivineMC start - MSPT Tracking for each world
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes60s = new MinecraftServer.TickTimes(1200);
+    // DivineMC end - MSPT Tracking for each world
+
     public ServerLevel(
         MinecraftServer server,
         Executor dispatcher,
