From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sun, 23 Feb 2025 01:14:54 +0300
Subject: [PATCH] Catch update suppressors


diff --git a/net/minecraft/network/PacketProcessor.java b/net/minecraft/network/PacketProcessor.java
index d501168e40f951ee37efcbc464c3ab1307051060..1b98de4bdf459646529c984b083e9988d1b2f3f5 100644
--- a/net/minecraft/network/PacketProcessor.java
+++ b/net/minecraft/network/PacketProcessor.java
@@ -97,6 +97,10 @@ public class PacketProcessor implements AutoCloseable {
             if (this.listener.shouldHandleMessage(this.packet)) {
                 try {
                     this.packet.handle(this.listener);
+                // DivineMC start - Catch update suppressors
+                } catch (org.bxteam.divinemc.util.exception.UpdateSuppressorException e) {
+                    LOGGER.info(e.getMessage());
+                // DivineMC end - Catch update suppressors
                 } catch (Exception var3) {
                     if (var3 instanceof ReportedException reportedException && reportedException.getCause() instanceof OutOfMemoryError) {
                         throw PacketUtils.makeReportedException(var3, this.packet, this.listener);
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index c5a1bf215aa91c2dc67926385fcb93f5e0cc3549..e6289059b20645c92b3e6b6b7d6da280e2f65fa5 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1823,6 +1823,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             serverLevel.tickTimes10s.add(this.tickCount, j);
             serverLevel.tickTimes60s.add(this.tickCount, j);
             // DivineMC end - MSPT Tracking for each world
+        // DivineMC start - Catch update suppressors
+        } catch (org.bxteam.divinemc.util.exception.UpdateSuppressorException e) {
+            LOGGER.info(e.getMessage());
+        // DivineMC end - Catch update suppressors
         } catch (Throwable levelTickingException) {
             CrashReport crashReport = CrashReport.forThrowable(levelTickingException, "Exception ticking world");
             serverLevel.fillReportDetails(crashReport);
diff --git a/net/minecraft/world/level/block/ShulkerBoxBlock.java b/net/minecraft/world/level/block/ShulkerBoxBlock.java
index 6c5696ab4981a6d582d4d0f13c9822bf84a5d9f1..95cbb9e988f4c54b37ff946e48e73e26436e9b37 100644
--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -181,7 +181,17 @@ public class ShulkerBoxBlock extends BaseEntityBlock {
 
     @Override
     protected int getAnalogOutputSignal(BlockState state, Level level, BlockPos pos, Direction direction) {
-        return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
+        // DivineMC start - Catch update suppressors
+        try {
+            return AbstractContainerMenu.getRedstoneSignalFromBlockEntity(level.getBlockEntity(pos));
+        } catch (ClassCastException ex) {
+            if (org.bxteam.divinemc.config.DivineConfig.FixesCategory.updateSuppressionCrashFix) {
+                throw new org.bxteam.divinemc.util.exception.UpdateSuppressorException(pos, this);
+            } else {
+                throw ex;
+            }
+        }
+        // DivineMC end - Catch update suppressors
     }
 
     public static Block getBlockByColor(@Nullable DyeColor color) {
diff --git a/net/minecraft/world/level/redstone/NeighborUpdater.java b/net/minecraft/world/level/redstone/NeighborUpdater.java
index 87512434ccd8d46d8db718228c790465c64b2c20..2928ed49be435d972683a3d8c68073c869c642c1 100644
--- a/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -60,9 +60,19 @@ public interface NeighborUpdater {
             state.handleNeighborChanged(level, pos, neighborBlock, orientation, movedByPiston);
             // Spigot start
         } catch (StackOverflowError ex) {
+            // DivineMC start - Catch update suppressors
+            if (org.bxteam.divinemc.config.DivineConfig.FixesCategory.updateSuppressionCrashFix) {
+                throw new org.bxteam.divinemc.util.exception.UpdateSuppressorException(pos, neighborBlock);
+            }
+            // DivineMC end - Catch update suppressors
             level.lastPhysicsProblem = pos.immutable();
             // Spigot end
         } catch (Throwable var9) {
+            // DivineMC start - Catch update suppressors
+            if (org.bxteam.divinemc.config.DivineConfig.FixesCategory.updateSuppressionCrashFix) {
+                throw new org.bxteam.divinemc.util.exception.UpdateSuppressorException(pos, neighborBlock);
+            }
+            // DivineMC end - Catch update suppressors
             CrashReport crashReport = CrashReport.forThrowable(var9, "Exception while updating neighbours");
             CrashReportCategory crashReportCategory = crashReport.addCategory("Block being updated");
             crashReportCategory.setDetail(
