--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -112,7 +_,7 @@
         .add(new ExplosionParticleInfo(ParticleTypes.POOF, 0.5F, 1.0F))
         .add(new ExplosionParticleInfo(ParticleTypes.SMOKE, 1.0F, 1.0F))
         .build();
-    public final List<TickingBlockEntity> blockEntityTickers = Lists.newArrayList();
+    public final org.bxteam.divinemc.util.BlockEntityTickersList blockEntityTickers = new org.bxteam.divinemc.util.BlockEntityTickersList(); // DivineMC - optimize block entity removals - Fix MC-117075
     protected final CollectingNeighborUpdater neighborUpdater;
     private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
     private boolean tickingBlockEntities;
@@ -1447,13 +_,11 @@
         boolean runsNormally = this.tickRateManager().runsNormally();
 
         int tickedEntities = 0; // Paper - rewrite chunk system
-        var toRemove = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<TickingBlockEntity>(); // Paper - Fix MC-117075; use removeAll
-        toRemove.add(null); // Paper - Fix MC-117075
         for (this.tileTickPosition = 0; this.tileTickPosition < this.blockEntityTickers.size(); this.tileTickPosition++) { // Paper - Disable tick limiters
             TickingBlockEntity tickingBlockEntity = this.blockEntityTickers.get(this.tileTickPosition);
             // Spigot end
             if (tickingBlockEntity.isRemoved()) {
-                toRemove.add(tickingBlockEntity); // Paper - Fix MC-117075; use removeAll
+                this.blockEntityTickers.markAsRemoved(this.tileTickPosition); // DivineMC - optimize block entity removals - Fix MC-117075
             } else if (runsNormally && this.shouldTickBlocksAt(tickingBlockEntity.getPos())) {
                 tickingBlockEntity.tick();
                 // Paper start - rewrite chunk system
@@ -1463,7 +_,7 @@
                 // Paper end - rewrite chunk system
             }
         }
-        this.blockEntityTickers.removeAll(toRemove); // Paper - Fix MC-117075
+        this.blockEntityTickers.removeMarkedEntries(); // DivineMC - optimize block entity removals - Fix MC-117075
 
         this.tickingBlockEntities = false;
         this.spigotConfig.currentPrimedTnt = 0; // Spigot
