--- a/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
@@ -331,6 +_,12 @@
                 if (!this.isSilent()) {
                     level.levelEvent(null, LevelEvent.SOUND_ZOMBIE_CONVERTED, this.blockPosition(), 0);
                 }
+
+                // DivineMC start - Fix MC-200418
+                if (org.bxteam.divinemc.config.DivineConfig.FixesCategory.fixMc200418 && mob.isPassenger() && mob.getVehicle() instanceof net.minecraft.world.entity.animal.chicken.Chicken && mob.isBaby()) {
+                    mob.removeVehicle();
+                }
+                // DivineMC end - Fix MC-200418
                 // CraftBukkit start
             }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.CURED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CURED // CraftBukkit
         );
